cmd/client.go

```diff
@@ -242,6 +252,129 @@ type Config struct {
 	UploadLimit       int64
 	DownloadLimit     int64
 	Transport         http.RoundTripper
+	// Force TLS
+	forceTLS bool
+}
+
+// getCredsChain returns an []credentials.Provider array for the config
+// and the STS configuration (if present)
+func (config *Config) getCredsChain() ([]credentials.Provider, *probe.Error) {
+	var credsChain []credentials.Provider
+	// if an STS endpoint is set, we will add that to the chain
+	if stsEndpoint := env.Get("MC_STS_ENDPOINT_"+config.Alias, ""); stsEndpoint != "" {
+		// set AWS_WEB_IDENTITY_TOKEN_FILE is MC_WEB_IDENTITY_TOKEN_FILE is set
+		if val := env.Get("MC_WEB_IDENTITY_TOKEN_FILE_"+config.Alias, ""); val != "" {
+			os.Setenv("AWS_WEB_IDENTITY_TOKEN_FILE", val)
+			if val := env.Get("MC_ROLE_ARN_"+config.Alias, ""); val != "" {
+				os.Setenv("AWS_ROLE_ARN", val)
+			}
+			if val := env.Get("MC_ROLE_SESSION_NAME_"+config.Alias, randString(32, rand.NewSource(time.Now().UnixNano()), "mc-session-name-")); val != "" {
+				os.Setenv("AWS_ROLE_SESSION_NAME", val)
+			}
+		}
+		stsEndpointURL, err := url.Parse(stsEndpoint)
+		if err != nil {
+			return nil, probe.NewError(fmt.Errorf("Error parsing sts endpoint: %w", err))
+		}
+		config.forceTLS = true
```

Comment:
I don't think it's a good idea to use a struct field to hide parameter transfer. Just pass it to `config.getTransport(forceTLS bool)` would be much cleaner. This may also introduce a bug, because you're also calling `config.getTransport()` in `cmd/client-admin.go`. Because you set `config.forceTLS` to `true` here, it will store it in the `config` and subsequent calls to `config.getTransport()` may always return TLS enabled transports.

Reply:
It wasn't in your previous code of the PR, so functionality has changed during this refactor. Suppose you have the following code (pseudo code):
```
cfg := Config{}
factory := NewAdminFactory(cfg)
transport1 := factory(cfg)     // may return a transport without TLS
_ = cfg.getCredsChain()
transport2 := factory(cfg)     // may return a transport with TLS
```
Because `Config.getCredsCain()` changes the internals of the `Config` structure, it will have side-effects that are unexpected. If you would pass `forceTLS` via parameters, then the `Config` structure will be untouched and `transport2` would have had the same result as the first call.

Even if this is the desired behaviour, I still consider it a bug because it's very unexpected that the order of these calls would be relevant.

Reply:
I think this code is hard to read and it doesn't need to be that way...

Reply:
PS: This will now force TLS always. I thought it should only enable TLS when `stsEndpointURL.Scheme == "https"`.